### PR TITLE
fix(turso): store and read back the same set of types (fixes #12628)

### LINKED ISSUE #12628: Turso writes every list value as [null], and nulls dictionary- and view-typed columns
## Summary

`scalar_value_to_turso` in `crates/data_components/src/turso.rs` is the only value converter on the
Turso accelerator's write path (`TursoDataSink::execute_insert_batch`). Three of its arms lose data
silently — the row is written, the load succeeds, and the column reads back `NULL` or `[null]`.

## 1. Every list value is written as `[null]`

`ScalarValue::List(Arc<ListArray>)` holds a **one-row** `ListArray` whose single row *is* the list.
The arm iterates that wrapper instead of the list's elements:

```rust
for i in 0..list_arr.len() {            // len() == 1 — the wrapper, not the elements
    let elem = ScalarValue::try_from_array(list_arr.as_ref(), i)?;   // → ScalarValue::List again
    match elem {
        ScalarValue::Int32(Some(v)) => ...,
        _ => json_values.push(serde_json::Value::Null),               // ← always taken
    }
}
```

`try_from_array` on a `ListArray` returns another `ScalarValue::List`, which falls to the catch-all,
so the output is `"[null]"` for **any** list, whatever its contents.

Confirmed against the code as it stands (`ScalarValue::List` over `Int32` `[1, 2, 3]`):

```
LIST     => Ok(Text("[null]"))
UTF8VIEW => Ok(Null)
DICT     => Ok(Null)
```

The read path (same file, `DataType::List` reconstruction) parses the stored text with
`serde_json::from_str::<Vec<i32>>`, which fails on `"[null]"` and appends a null list — so
`List<Int32>`, the *one* element type the reader supports, round-trips to `NULL` for every row.

Two smaller defects sit in the same arm: the element `match` accepts only `Int32`/`Int64`/`Utf8`
and nulls everything else, and `LargeList`/`FixedSizeList` never reach it at all — they fall to the
function's final `_ => TursoValue::Null`, even though the accelerator's DDL declares all three as
`TEXT` (`crates/runtime/src/dataaccelerator/turso.rs`, `DataType::List(_) | DataType::LargeList(_) |
DataType::FixedSizeList(_, _) => "TEXT"`).

## 2. `Utf8View` / `BinaryView` values are written as NULL

Neither variant has an arm, so both hit `_ => TursoValue::Null`. The DDL maps them to `TEXT`/`BLOB`
through its own `_` default, so the column exists and is silently filled with NULLs. Sibling write
paths in this repo already handle them — `snowflake/write.rs` and `delta_lake.rs` both match
`ScalarValue::Utf8View(Some(v))` alongside `Utf8`/`LargeUtf8`.

## 3. Dictionary-encoded values are written as NULL

`ScalarValue::Dictionary(_, inner)` has no arm either, so a dictionary-encoded column — what the
MySQL and PostgreSQL replication paths in `crates/data_components` produce, among others — writes
NULL for every row. `snowflake/write.rs` has the one-line precedent:
`ScalarValue::Dictionary(_, value) => scalar_to_sql_literal(*value, column)`.

## Root cause common to all three

The function ends in `_ => TursoValue::Null`, so an unhandled type is indistinguishable from a real
SQL NULL. Any variant added to `ScalarValue` upstream, or simply forgotten here, degrades to silent
data loss rather than a load failure — which is the inverse of the repo's stated rule that a
structured error beats possibly-wrong data.

## Impact

Silent, permanent data loss in an accelerated table: the affected columns are readable and contain
NULL, with nothing in the logs. Affects any dataset accelerated with `engine: turso` that has a
list, dictionary-encoded, or view-typed column.

## Affected versions

`trunk` and `release/2.1`.

## Proposed fix

- Serialize a list from the element values (`list_arr.value(0)`), and route `LargeList` /
  `FixedSizeList` through the same helper.
- Cover the scalar element types JSON represents exactly (integers, floats, booleans, strings) and
  reject the rest explicitly instead of writing `null`.
- Add `Utf8View` / `BinaryView` arms, and unwrap `Dictionary(_, inner)` recursively.
- Replace the trailing `_ => TursoValue::Null` with an explicit null test plus a typed error naming
  the unsupported type, so a future gap fails the load instead of nulling the column.

### LINKED ISSUE #12631: Turso reads back fewer types than it stores, so a list, map or dictionary column fails its scan
## Summary

`TursoTableProvider::values_to_record_batch` reconstructs a narrower set of Arrow types than the
Turso accelerator's DDL will declare and its write path will store. A column of one of the missing
types fails its scan with an Arrow schema mismatch — the reader builds an array of a different type
than the schema's field, and `RecordBatch::try_new` rejects it.

Measured against the read path as it stands, with a stored value present for the column:

```
List<Float64>    => expected List(Float64) but found List(Int32)
List<Utf8>       => expected List(Utf8) but found List(Int32)
LargeList<Int32> => expected LargeList(Int32) but found Utf8
Dictionary       => expected Dictionary(Int32, Utf8) but found Utf8
```

The `_` arm at the end of the match builds a `StringArray` for anything it does not recognize, which
only matches a `Utf8` field, so every unrecognized type lands the same way.

## Missing types

- **Lists** — only `List` with `Int32` elements is reconstructed. Any other element type produces a
  `List<Int32>` of nulls, and `LargeList` / `FixedSizeList` are not handled at all.
- **Maps** — only `Utf8` keys to `Int32` values. The value is also read with `as_i64() as i32`, which
  truncates rather than failing.
- **`Dictionary`, `Float16`, `FixedSizeBinary`** — no arm.

## Why now

#12628 fixed the write path, which had been storing `"[null]"` for every list and `NULL` for
dictionary- and view-typed values. Its PR also added `Utf8View` and `BinaryView` to the read path, so
those two round-trip. The rest remain write-only: the value is now stored faithfully, but nothing can
read it back.

The asymmetry predates #12628 — these columns failed to scan before it too, because the stored value
was wrong *and* unreadable. What is new is that the stored bytes are now correct, so closing this
gap makes existing accelerated data readable without a reload.

## Proposed fix

Reconstruct each type the write path stores:

- Parse a JSON list into the element type the field declares, for `List`, `LargeList` and
  `FixedSizeList`.
- Parse a JSON object into the map's declared key and value types, and fail rather than truncate a
  value that does not fit.
- Rebuild a `Dictionary` column from its inner value type, and add `Float16` / `FixedSizeBinary`.
- Replace the `_ => StringArray` fallback with an error naming the type, so a future gap is a clear
  failure rather than a schema mismatch reported against an unrelated column.

### PR DESCRIPTION
**Pull request by claudespice**
## Summary

`scalar_value_to_turso` is the only value converter on the Turso accelerator's write path, and three
of its arms lost data without failing the write. The row was inserted, the load reported success,
and the column read back `NULL`.

**A list scalar wraps its elements in a one-row array, and the arm iterated the wrapper.**
`ScalarValue::List(Arc<ListArray>)` holds a single row whose value *is* the list, so `list_arr.len()`
is `1` and `try_from_array(list_arr, 0)` hands back another `ScalarValue::List` — which fell to the
element match's catch-all. Every list, of every element type, serialized as `"[null]"`. The read
path then parsed that with `serde_json::from_str::<Vec<i32>>`, failed, and appended a null list, so
`List<Int32>` — the one element type the reader supports — round-tripped to `NULL` for every row.

**`Utf8View`, `BinaryView` and dictionary-encoded values had no arm at all**, so they reached the
trailing `_ => TursoValue::Null` and wrote NULL. The DDL declares these columns through its own
`_ => "TEXT"` default, so the column exists and quietly fills with nulls. Both sibling write paths in
this repo already handle them: `snowflake/write.rs` and `delta_lake.rs` match `Utf8View` alongside
`Utf8`/`LargeUtf8`, and `snowflake/write.rs` unwraps `Dictionary(_, value)` recursively.

**A map scalar sliced out of a batch keeps the whole entries child array.** `MapArray::keys()` and
`values()` return the children of the array the scalar was sliced from, not the row's own entries, so
a map column serialized every row's entries into every row.

The shared root cause is the trailing `_ => TursoValue::Null`: an unhandled type was indistinguishable
from a real SQL NULL, so a gap in coverage degraded to silent data loss instead of a load failure.

## Changes

- Serialize a list from its element values (`value(0)`), and route `LargeList` and `FixedSizeList`
  through the same helper — the accelerator's DDL declares all three as `TEXT`.
- Add a `scalar_value_to_json` helper covering the types JSON represents exactly (integers, floats,
  booleans, strings, and dictionaries of those), so a list element is no longer limited to
  `Int32`/`Int64`/`Utf8`. A non-finite float is rejected rather than encoded as `null`, which
  `serde_json` would otherwise make indistinguishable from a null element.
- Read a map's entries through `value(0)`, keep an entry whose value is null instead of dropping it,
  and reject a non-string key instead of silently discarding the entry.
- Add `Utf8View`, `BinaryView`, `FixedSizeBinary` and `Float16` arms, and unwrap
  `Dictionary(_, inner)` recursively.
- Test `is_null()` once up front so a typed null of any variant still stores as SQL NULL, then
  replace the trailing catch-all with an error naming the unsupported type.
- Read the stored list as `Vec<Option<i32>>`, so a list holding a null comes back as a list holding a
  null rather than as a null list.
- Reconstruct `Utf8View` and `BinaryView` on the read path, so the two view types the write path now
  stores actually round-trip rather than failing their scan on a schema mismatch.

### Behavior change

A column whose type the accelerator cannot store now fails the write instead of loading as NULLs —
`Struct` is the practical case. This is deliberate: the read path has no `Struct` reconstruction
either (it falls through to a `StringArray` that does not match the declared field), so such a column
was never usable, and a load that stops with a typed error naming the type is strictly better than
one that reports success over an empty column.

### Read support is still narrower than write support

`values_to_record_batch` reconstructs a list only with `Int32` elements, a map only from `Utf8` to
`Int32`, and neither `LargeList`, `FixedSizeList`, `Dictionary`, `Float16` nor `FixedSizeBinary` at
all. Those columns fail their scan with an Arrow schema mismatch — measured against the read path as
it stands, and true whatever the writer stored:

```
List<Float64>    => expected List(Float64) but found List(Int32)
LargeList<Int32> => expected LargeList(Int32) but found Utf8
Dictionary       => expected Dictionary(Int32, Utf8) but found Utf8
```

So the asymmetry predates this change: those columns were already unscannable, and the writer was
also corrupting them. Narrowing the writer to the readable subset instead was considered and
rejected: it would fail the whole dataset's load over a column that projection pushdown may never
touch, trading a scannable dataset for none at all.

# The read path (#12631)

The reader is now the inverse of the writer, so there is no longer a set of types that store without
error and cannot be read back. It had rebuilt a narrower set than the DDL declares:

| Declared | Read back as |
|---|---|
| `List<Float64>`, `List<Utf8>`, any non-`Int32` element | `List<Int32>` of nulls |
| `LargeList`, `FixedSizeList` | `Utf8` (the string fallback) |
| `Dictionary`, `Float16`, `FixedSizeBinary` | `Utf8` (the string fallback) |
| `Map<_, Int32>` value past `i32` | a different, valid-looking number |

Each element is now decoded as the type its field declares. Four decisions worth calling out, since
two of them depart from what the issue proposed:

- **A value that does not fit nulls the whole cell, not the element.** #12631 asked for a hard error
  instead of the `as_i64() as i32` truncation. This does reject the truncation — but as a soft
  failure, because that is this function's documented contract (*"values that cannot be converted to
  the target type are converted to NULL … prioritizes query availability"*) and because a null cell
  over a non-NULL stored value is exactly what the existing `count_conversion_failures` counts and
  WARNs about. A dropped *element* inside a valid list would be invisible to it, which is why the
  cell rather than the element is what fails. Say so if you would rather it hard-error.
- **One exception, and it fails loudly.** A `Float16` element is decoded as `Float64` and narrowed by
  Arrow's cast with `safe: false`, because `half::f16` is not a dependency of this crate. An element
  outside `f16`'s range therefore fails the scan rather than nulling the cell. It is unreachable from
  data this writer produced (the value came *from* an `f16`), and the alternative — `safe: true` —
  would silently null the element, which nothing counts. Taking `half` as a direct dependency would
  remove the special case entirely; that seemed like more than this bug warrants, and is easy to do
  later.
- **A `Map`'s `sorted` flag is preserved.** The old arm discarded it.
- **A `FixedSizeList` of the wrong width reads as null**, in both directions — padding a short list
  invents elements and truncating a long one drops them.

The per-column match moves into `column_from_turso_values` so the `Dictionary` arm can recurse into
its value type, which is how the writer stores it. **Apart from the new and rewritten arms that move
is whitespace-only — `git diff -w` shows it** (1153/625 insertions/deletions become 689/161).

## Performance impact

The generic element decode goes through `ScalarValue` rather than a typed Arrow builder, so a list or
map column costs one boxed `ScalarValue` per element where the old `List<Int32>` path used
`Int32Builder` directly. That is the price of being generic over the declared element type without a
per-type builder arm for every list flavour, and it applies only to list/map/dictionary columns —
every scalar column keeps its existing typed path untouched. Correctness over throughput here: the
alternative was continuing to return nulls, wrong numbers, or a schema mismatch.

The `_ => StringArray` fallback now names the type it could not read. It only ever deferred the
failure to `RecordBatch::try_new`, which reported it as a mismatch against whichever column Arrow
compared first — so this changes the message, not whether the scan fails.

Raised by the adversarial review, along with one deferral: a `mode: file` accelerator written before
this fix holds `"[null]"` payloads that the relaxed reader now accepts as a one-element null list
rather than flagging as a conversion failure. Distinguishing them needs an encoding version and a
forced rebuild, which is a migration mechanism rather than part of this repair — filed as #12632.
Any refresh clears the condition.

## Test plan

- 24 new unit tests in `crates/data_components/src/turso.rs`, including an end-to-end pass through
  `TursoDataSink::write_all` into an in-memory Turso database and, for the read path, a
  `assert_round_trips` helper that stores a value through `scalar_value_to_turso` and reads it back
  through `values_to_record_batch`, asserting the array that comes out equals the one that went in —
  the symmetry property this PR is about, rather than a hand-written JSON fixture.
- 503 `data_components` lib tests pass, re-run after merging the base branch.
- **Neuter matrix, 17/17 caught.** The write-path twelve as before. Five more for the read path,
  each caught by the test named for it: make an integer element wrap instead of nulling the cell;
  decode only `Int32` elements; reshape a wrong-width `FixedSizeList` into the cell; restore the
  silent string fallback; and — as a control against the plausible-but-wrong fix — substitute a null
  map *entry* for an undecodable value instead of nulling the cell, which only the
  `count_conversion_failures` assertions can see.
- Worth recording that the wrong-width neuter **passed on the first run**, because the test only
  covered a list that was too *long*. The gap was real, not the neuter's: the test now covers both
  directions, and the neuter fails.
- `make lint`
- `make test`

## Checklist

- [x] Regression test fails on the old code and passes on the new
- [x] Null, empty, and boundary cases covered (null list, null element, empty list, typed nulls,
      non-finite float, non-string map key)
- [x] No `.unwrap()` / `.expect()` in production code
- [x] Error messages name the type, never the value
- [x] `make lint`
- [x] `make test`

## Sign-off status

Not dispatched deliberately, and the red `Attestation` is not about this diff. `make nextest` runs
`cayenne::mutation_property_test prop_sequential_memory_impl_turso`, which overflows a 2 MiB stack on
every branch (#12436) and is not in the Makefile's by-name exclusion list. The fix is #12561, which
is approved and queued but has not landed. Dispatching now would spend ~3.5h to post
`signoff=failure`, which *disqualifies* the commit — strictly worse than leaving it pending. Worth
re-running once #12561 merges.

Fixes #12628
Fixes #12631

### CODE REVIEW

**Comment by claudespice** [comment]
@copilot review

File: crates/data_components/src/turso.rs [hunk 1]
```diff
@@ -676,6 +676,17 @@ impl TursoTableProvider {
...
+                    let values: Vec<Option<String>> = rows
+                        .iter()
+                        .map(|row| match &row[col_idx] {
+                            TursoValue::Text(s) => Some(s.clone()),
+                            TursoValue::Null => None,
+                            _ => None,
+                        })
+                        .collect();
+                    Arc::new(StringViewArray::from(values))
+                }
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/data_components/src/turso.rs]: Applied in `cb05761cc` — the arm now collects `Vec<Option<&str>>` via `s.as_str()`.

Two notes on the reasoning, since one half of it does not hold:

The allocation is real. `Vec<Option<String>>` heap-allocates a copy of every non-null value, and `StringViewArray` then copies it again into the view buffer, so the first copy buys nothing. Confirmed the borrowed form is available on the arrow rev this workspace pins: `impl From<Vec<Option<&str>>> for StringViewArray` at `arrow-array/src/array/byte_view_array.rs:1164`.

"Largely defeats the benefit of using view arrays" overstates it. A `Utf8View` column is built here because that is the type the schema asks for, and what the view layout buys downstream — inline short strings, cheap slicing and comparison — is unaffected by how the array was constructed. The cost was a construction-time waste, not a lost property of the array.

The clearest argument for the change is one the diff already contained: the `BinaryView` arm two match arms below was borrowing (`Vec<Option<&[u8]>>`, `b.as_slice()`) while this one cloned. The two new view columns now build the same way.

Verified on the new head: `cargo check`/`clippy -p data_components --features turso --lib --tests` clean, `cargo fmt --check` clean, `cargo test -p data_components --features turso --lib turso` 34 passed / 0 failed.

**Comment by claudespice** [comment]
### Sign-off state, so the next pass does not re-derive it

The `signoff=failure` currently on `78a07f63a` is **from a cancelled run, not a gate result**. Detail
of the four attempts on this branch:

| run | outcome | what actually happened |
|---|---|---|
| [31091971484](https://github.com/spiceai/spiceai/actions/runs/31091971484) | cancelled | a concurrent dispatch by PR number (`Remote Sign-off (PR #12633)`) shared this branch's concurrency group and evicted it after 44 min queued + 95 min running |
| [31101325211](https://github.com/spiceai/spiceai/actions/runs/31101325211) | failure | **real, and fixed.** `test_scalar_value_to_turso_keeps_a_null_map_value` asserted a serialized JSON object's key order; `crates/runtime` enables `serde_json/preserve_order` and cargo unifies it across the workspace, so `serde_json::Map` is `IndexMap` under `make nextest` and `BTreeMap` under a scoped `cargo test -p data_components`. Fixed in `78a07f63a` by comparing parsed values, verified with `preserve_order` both on and off |
| [31117872937](https://github.com/spiceai/spiceai/actions/runs/31117872937) | cancelled | `##[error]The operation was canceled` with **zero** `-->` markers anywhere in the log — no rustc diagnostic. The `signoff=failure` status now on the head is this run's cancellation handler firing, not a gate verdict |
| [31120390127](https://github.com/spiceai/spiceai/actions/runs/31120390127) | failure | the sign-off job recorded **0 steps** over 10 minutes — assigned to a runner that died before executing anything |

So the gate has run to completion on this branch exactly once, and the only genuine failure it found is
already fixed. The last two attempts never executed it.

The same run that found the real failure also aborted on a test unrelated to this change:

```
SIGABRT cayenne::mutation_property_test prop_sequential_memory_impl_turso
    thread has overflowed its stack
```

`scalar_value_to_turso` is a private `fn` in `data_components`, so `cayenne` cannot reach it. It is a
fifth member of the family tracked by #12436 whose `Makefile` exclusion names only four — reported
[there](https://github.com/spiceai/spiceai/issues/12436).

Not re-dispatching further this pass: five sibling sign-off runs are queued ahead, and a fourth
attempt would evict someone else's without testing anything new.

**What a reviewer needs to know:** the diff itself is verified locally — 34 turso tests (14 new), the
full 490-test `data_components` lib suite, scoped clippy clean, and a 12/12 neuter matrix where each
neuter is caught by the test named for it.

**Comment by grokspice** [comment]
### The `signoff=failure` on `78a07f63a` is not this diff — do not patch or re-dispatch yet



Run [31134355008](https://github.com/spiceai/spiceai/actions/runs/31134355008) ran 13291s and ended
`Error 100`. One test failed out of 9789:

```
SIGABRT [0.284s] (9785/9789) cayenne::mutation_property_test prop_sequential_memory_impl_turso
  thread 'prop_sequential_memory_impl_turso' has overflowed its stack
  fatal runtime error: stack overflow, aborting
Summary [1010.388s] 9789 tests run: 9788 passed, 1 failed, 27 skipped
```

That is the known cayenne/turso stack overflow, and **#12561** ("run a backend-parameterized test
on a stack it fits") is the fix. #12561 is approved and sitting in the merge queue, so this clears
itself once it lands.

One thing that is *not* a failure and shouldn't be chased: `runtime-async
tests::test_managed_tokio_runtime_handle` shows `TRY 1 FAIL` on the assertion
`!join_handle.is_finished()` at `crates/runtime-async/src/lib.rs:235`, then passes on retry 2 —
nextest reports it as `1 flaky`, not as the failure. The run's only hard failure is the SIGABRT.

**Next pass:** wait for #12561 to merge, then re-dispatch sign-off against the then-current HEAD.
Re-dispatching before that just spends another ~3.5h reaching the same abort.

**Review by lukekim — APPROVED** [review]
